### ISSUE TITLE
groupby.nth lost multiindex

### ISSUE THREAD

**Issue by jesrael** [pydata/pandas#11830]
It is bug or not? Because in function `mean` and `first` it is OK.
[link](http://stackoverflow.com/questions/34237462/selecting-column-before-nth-breaks-group-indices/)

    df = pd.DataFrame({'a': [1, 1, 2, 2], 'b': ['b', 'b', 'b', 'a'], 'c': [1, 2, 3, 4]})
    print df
    #   a  b  c
    #0  1  b  1
    #1  1  b  2
    #2  2  b  3
    #3  2  a  4
    
    #lost multiindex
    print df.groupby(['a', 'b']).c.nth(0)
    #0    1
    #2    3
    #3    4
    #Name: c, dtype: int64
    
    print df.groupby(['a', 'b']).c.mean()
    #a  b
    #1  b    1.5
    #2  a    4.0
    #   b    3.0
    #Name: c, dtype: float64
    print df.groupby(['a', 'b']).c.first()
    #a  b
    #1  b    1
    #2  a    4
    #   b    3
    #Name: c, dtype: int64
    print df.groupby(['a', 'b']).nth(0).c
    #a  b
    #1  b    1
    #2  a    4
    #   b    3
    #Name: c, dtype: int64

**Comment by jreback** [comment]
looks like a bug, but this is a quite complicated area, see [here](https://github.com/pydata/pandas/issues/6944)

If you'd like to dig-in would be great.

**Comment by pwaller** [comment]
By the way, I just hit this I think, and you don't need a multi-index to cause it:

```
df.groupby(df.b).a.nth(0), x.groupby(df.b).a.first()
```

gives:

```
(0    1
 3    2
 Name: a, dtype: int64,
 b
 a    2
 b    1
 Name: a, dtype: int64)
```

(And I would expect the two to be equal). It seems that for whatever reason nth is having its index lost.

**Comment by pwaller** [comment]
I just built several versions in the past to see if this was a regression. It's doesn't look like a regression. I've tested as far back as 0.12.

**Comment by pwaller** [comment]
Sorry. I think my test may have been depending on a feature which didn't exist in older versions (the `.name` property of indices). It now looks like it gives the correct result in v0.13.0. This is my test:

```
$ python3 -c 'import pandas as P; raise SystemExit(not P.DataFrame({"a": [1, 2, 3, 4], "b": [9, 9, 9, 9]}).groupby("a").a.nth(0).index.equals([1, 2, 3, 4]))'
$ echo $?
0
```

The latest version doesn't manifest this.

Going to try an automated git bisect. Any hints on how to make pandas build faster in development mode?

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/groupby.html#groupby-nth

``nth`` actually has slightly different semantics than ``.first``. we have been discussing this in a couple of issues: #7569, #11038, and #11039

**Comment by pwaller** [comment]
OK. Not that it matters but I was able to bisect the behaviour change to c444c73c23691da76bd09365db78b637a216c6ea.

FWIW, For me this wasn't very clearly explained in the documentation, for example [Here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.GroupBy.nth.html#pandas.core.groupby.GroupBy.nth). I guess it comes down to how you read the word "rows". For me, that just meant that it was picking out a row within the group, but I still expected it to have the index for that group - just as first() and last() do. It would be very useful if the docs made this behaviour difference clear.

Another thing that is a surprise to me:

```python
P.DataFrame({"a": [1, 2, 3, 4], "b": [9, 9, 9, 9]}).groupby("a").a.nth(0)
```

Is different from:

**Comment by pwaller** [comment]
Note to anyone following along by email: I [edited the previous comment](https://github.com/pydata/pandas/issues/11830#issuecomment-170285882) after accidentally sending it prematurely, and I will delete this comment from the tracker.

**Comment by jreback** [comment]
the issue is we need a bunch more test as their is s bug in the results - can u do a PR with some more tests - would help move this along

**Comment by adneu** [comment]
I believe this was fixed with #11039. That commit now makes `Groupby.nth` reducing for a `Series` (keeping the original index) vs. the former behavior, which was filtering (position based index). Also that commit added these tests which I think address this bug.
```
+        assert_series_equal(g.B.nth(0), df.set_index('A').B.iloc[[0, 2]])
+        assert_series_equal(g.B.nth(1), df.set_index('A').B.iloc[[1]])`
```
Do you think more tests would be useful?

**Comment by jreback** [comment]
ok #12839 closed this, but see the comment on the actual PR (we ended up re-opening that one) because the state was not being preserved.

welcome to have you address that issue.

thanks!
